content/blog/2022/2022-09-25-node-urlShortener.md

```diff
@@ -0,0 +1,494 @@
+
+These generated shortcodes are stored alongside the domain name.
+
+And then mapped to the long URL.
+
+When we visit the newly generated short URL. A database lookup is performed using the shortcode as a key, and if it is found, we are redirected to the long URL's content.
+
+URL shortener service simply uses a URL redirect. However, scaling, monitoring, URL filtering, spam prevention, URL verification, and other required features can result in more advanced logics.
+
+## Setting up `Node.Js` Application
+
+To begin, move into a new root directory where we want our application to live.
+
+Enter the following command into the terminal:
+```
+mkdir urlbackend
+```
+And this:
+```
+cd urlbackend
+```
+The folder `urlbackend` is created here. Then we moved our terminal directory to `urlbackend`, which is where our `Node.js` server lives.
+
+Run the command below in the terminal to initialize our `Node.js` application.
+```
+npm init -y
+```
+Then open your `Node.js` application in your preferred IDE. 
+
+Using the terminal run the following command. Ensure you are in the `urlbackend` directory:
+
+```
+npm install cors dotenv express mongoose shortid
+```
+Here, we're setting up the required dependencies for our backend, which include:
+- cors: Cross-origin resource sharing (CORS) allows AJAX requests to skip the Same-origin policy and access resources from remote hosts. Comes in handy while connecting to our react application.
+- dotenv: Loads environment variables from a .env file into process.env
+- express: `Node.Js` framework
+- mongoose: An object modeling tool aids in connecting to and communicating with `MongoDB`.
+- shortid: Generates non-sequential short unique ids
+
+Next, create an `index.js` and a  `.env` file.
+
+Our application should be structured like this:
+{{% image alt="We want to deploy our application into a staging and a production environment" src="images/posts/designing-a-aws-cdk-project/serverapp.png" %}}
+
+In the `index.js` file, paste this to create a simple`Node.Js` server:
+```javascript
+const express = require('express');
+const app = express();
+
+// Server Setup
+const PORT = process.env.PORT || 3333;
+app.listen(PORT, () => {
+  console.log(`Server is running at PORT: ${PORT}`);
+});
+```
+In the code above, we created an express server by importing and instantiating the express package into the index.js file. Making it listen on our custom PORT `3333`.
+
+Then next in our `.env` file we are going to store our `DOMAIN_URL` and `MongoDB_URI`. For the time being, the `DOMAIN_URL` in this application will be our `localhost` server location. 
+
+To get our `MongoDB_URI` we have to create a MongoDB atlas cluster
+
+## Creating MongoDB Atlas Cluster
+Follow these steps to create a MongoDB altas cluster:
+- [Go here](https://account.mongodb.com/account/register)  to sign up for a new MongoDB Atlas account.
```

Comment:
Why are we creating a MongoDB Atlas Cluster? What is that even? Why do we need a database at all? Why do we need a cluster?